include/onnxruntime/core/providers/nnapi/nnapi_provider_factory.h

```diff
@@ -45,9 +45,13 @@ enum NNAPIFlags {
   // and NNAPI_FLAG_CPU_ONLY flags are set
   NNAPI_FLAG_CPU_ONLY = 0x008,
 
+  // DISABLE_CPU may bring perf improvement, but may also make it initialize failed.
+  // A soft disabling flag is used to try disable CPU first, and if it fails, fallback to CPU again.
+  NNAPI_FLAG_CPU_DISABLED_SOFT = 0x010,
```

Comment:
would be good to clearly indicate when a user would want to use this option. e.g., why would this be better than not specifying CPU_DISABLED?

also, how does this interact with the other options? I guess we would want it to be mutually exclusive with CPU_ONLY and CPU_DISABLED.

Reply:
Not follow "this option interact with the others". Could you elaborate more?

Reply:
Trying to understand how this setting provides the user with a meaningful option.

- For newer NNAPI version we can query individual devices and prevent NNAPI using its CPU implementation by setting the feature level to the max non-CPU device
  - assumption: if the device says it implements a feature level of NNAPI it implements everything in that level, so as long as we limit things to the max non-CPU device that means only non-CPU devices will be used by NNAPI

- For older NNAPI versions we can't do that so the slow NNAPI reference CPU implementation potentially comes into play. 
  - we never want to use that as the performance is generally terrible as it's not optimized in any way, and our CPU kernels are much better
    - which possibly makes the choice between NNAPI on or off, as using NNAPI with CPU fallback is expected to have bad performance and not be a viable option

Given that, in what scenario would a user choose NNAPI_FLAG_CPU_DISABLED_SOFT and expect better performance than if they disabled NNAPI?

Reply:
> Sorry, not quite follow "this option interact with the others". Could you elaborate more?

sure. for example, what happens if the new CPU_DISABLED_SOFT flag is specified along with either CPU_ONLY or CPU_DISABLED? or should that be disallowed?

Reply:
> Trying to understand how this setting provides the user with a meaningful option.
> 
> * For newer NNAPI version we can query individual devices and prevent NNAPI using its CPU implementation by setting the feature level to the max non-CPU device
>   
>   * assumption: if the device says it implements a feature level of NNAPI it implements everything in that level, so as long as we limit things to the max non-CPU device that means only non-CPU devices will be used by NNAPI
> * For older NNAPI versions we can't do that so the slow NNAPI reference CPU implementation potentially comes into play.
>   
>   * we never want to use that as the performance is generally terrible as it's not optimized in any way, and our CPU kernels are much better
>     
>     * which possibly makes the choice between NNAPI on or off, as using NNAPI with CPU fallback is expected to have bad performance and not be a viable option
> 
> Given that, in what scenario would a user choose NNAPI_FLAG_CPU_DISABLED_SOFT and expect better performance than if they disabled NNAPI?

I think this flag is a bridge and  backing out for NNAPI_FLAG_CPU_DISABLED. If we set this flag, NNAPI ep either gives a better performance or just failed during initialization.

In contrast, If user set NNAPI_FLAG_CPU_DISABLED_SOFT, NNAPI EP can run model at a higher probability and banlance performance at the same time.

The scenario we designed is to replace NNAPI_FLAG_CPU_DISABLED at some extent.

Reply:
you are right, this flag can't  provide any further info since we have gotten the best feature level whenever nnapi-CPU is enabled or not.  So It doesn't provide better perf.